File: leetcode/114/P114.rb
# 114. Flatten Binary Tree to Linked List
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def flatten(root)
	search root if root
end

def search node
	tail, right = node, node.right
	if node.left
		node.right = node.left
		tail = search node.left
		node.left = nil
	end
	if right
		tail.right = right
		tail = search right
	end
	tail
end

File: leetcode/114/src/P114.java
/**
 * Created by qwordy on 2/26/16.
 * 114. Flatten Binary Tree to Linked List
 */

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  public void flatten(TreeNode root) {
    if (root != null) search(root);
  }

  // Return the last node
  private TreeNode search(TreeNode node) {
    TreeNode tail = node;
    TreeNode right = node.right;
    if (node.left != null) {
      node.right = node.left;
      tail = search(node.left);
      node.left = null;
    }
    if (right != null) {
      tail.right = right;
      tail = search(right);
    }
    return tail;
  }
}

public class P114 {
}
